src/Datadog.Trace/Util/DomainMetadata.cs

```diff
@@ -100,12 +101,15 @@ public static int AppDomainId
 
         public static bool ShouldAvoidAppDomain()
         {
-            if (AppDomainName.IndexOf("ApplicationInsights", StringComparison.OrdinalIgnoreCase) >= 0)
+            var appDomain = AppDomain.CurrentDomain;
+
+            if (!(appDomain.GetData(IsAppInsightKey) is bool isAppInsight))
             {
-                return true;
+                isAppInsight = AppDomainName.IndexOf("ApplicationInsights", StringComparison.OrdinalIgnoreCase) >= 0;
+                appDomain.SetData(IsAppInsightKey, isAppInsight);
```

Comment:
Since we only use the current appdomain, I'm wondering if we can go even further and cache the value in a static field. With all those "domain neutral" thingies, I'm not sure anymore of whether this instance is shared across appdomains or not

Reply:
Static fields are not shared across AppDomains (even if the assembly is domain-neutral) so that's a good idea. There's some good information here on how domain-neutral assemblies are treated: https://docs.microsoft.com/en-us/dotnet/framework/app-domains/application-domains

Reply:
Accessing static fields is _slower_ if the assembly is shared. Since all the AppDomains share the same code, but the field value is not shared, static field access requires an extra level of indirection. If this is a value we will be reading a lot in domain-neutral scenarios, making it static could be counter-productive.

On the other hand, I don't think we are every loaded domain-neutral in AAS, so if this value is only read in AAS, it's probably fine.

Reply:
`GetData` does a bunch of checks, acquires a lock, then fetches the boxed value from a dictionary. My gut feeling is that whatever mechanism is involved in fetching static field in domain-neutral assemblies, it shouldn't be slower than that. Unfortunately, the assumption seems really hard to test :/